### ISSUE TITLE
percentage of minsize/maxsize

### ISSUE THREAD

**Issue by fred-wang** [w3c/mathml-core#103]
From MathML Core: "Percentages value for minsize and maxsize properties of an embellished operator are interpreted relative to the target stretch size before application of size constraints, as described in § 3.3.1.1 Algorithm for stretching operators along the block axis. "

This is not really clear.

The current version of the spec says "Let minsize and maxsize be the minsize and maxsize properties on the operator. Percentage values are intepreted relative to T = Tascent + Tdescent"

So it should really refer to section https://w3c.github.io/mathml-core/#layout-of-operators

**Comment by fred-wang** [comment]
Additionally, default minsize of "100%" does not work with the current definition as this overrides maxsize. So I guess it should instead be 1em or 0.

**Comment by brucemiller** [comment]
Sorry I missed this when it went by before, but "relative to the target stretch size" is in marked contrast to MathML Full's "multipliers of the operator's normal size" (ie. the unstretched size; See https://w3c.github.io/mathml/spec.html#presm_op_stretch).

Was this an intentional change?  Should this issue be reopened, or a new one?

**Comment by dginev** [comment]
@fred-wang I second Bruce's question.

Also, when the behavior is finalized, I would also like to request an example in the MathML Core text that walks through the steps to compute the absolute size when starting with the percentage value of maxsize/minsize.

The current text is a bit hard to follow for me (and it wasn't immediately clear if it applies to all `<mo>` cases or only "If the operator has the stretchy property").

Here are two minimal examples that helped me appreciate the nuances with the containing row, though they don't render quite right yet - but maybe that is an independent issue:

https://gist.github.com/dginev/d9611039683098aba904fddd750acc16

**Comment by brucemiller** [comment]
It would be good if a test verified that `symmetric=true` is respected when `minsize` and/or `maxsize` are given.

**Comment by dginev** [comment]
I have been asked by arXiv to raise awareness of the impact of this issue, cross-referencing https://github.com/arXiv/html_feedback/issues/354#issuecomment-1962385739

**Comment by bkardell** [comment]
<img width="1230" alt="Screenshot 2024-02-27 at 11 00 10 AM" src="https://github.com/w3c/mathml-core/assets/870501/7f9a74c3-8502-41d2-818e-071ec25693de">

I love the pens for people to try, but also some might not be able to check certain browsers easily and the results can change over time which can make it confusing if the issue is open very long, so ... Below is a screen grab of @dginev's pen opened in 3 browsers side by side. From left to right: Firefox, Chrome, Safari

<img width="1230" alt="Screenshot 2024-02-27 at 11 00 10 AM" src="https://github.com/w3c/mathml-core/assets/870501/cb7da8a6-9008-498b-ba1e-306cbbc16ebb">

**Comment by brucemiller** [comment]
Given the several browser quirks, particularly the unsymmetric "|", 90% may be a bit too subtle to appreciate the effect of minsize, maxsize. This more artificial example shows Firefox (left) tracking MathML Full; Chrome (middle) tracking MathML Core and Safari (left) apparently ignoring percentages in minsize, maxsize:
![percent](https://github.com/w3c/mathml-core/assets/2972564/ea36e2bb-7884-4785-99f5-4105ce2dff66)

**Comment by brucemiller** [comment]
Core currently specifies that percentages in minsize/maxsize are  "Percentage values are interpreted relative to T = Tascent + Tdescent." where T is apparently the target, ie. stretched, size of the glyph.

This is an incompatible change from MathML 2 & 3 (at least). Although the current MathML Full spec sometimes refers simply to percent as relative to "normal" size, which apparently allows for some (mis)interpretation), in most other places it is explicit.
See https://w3c.github.io/mathml/spec.html#dictionary-based-attributes
"Unitless or percentage values indicate a multiple of the reference size, being the size of the unstretched glyph."

Consequently we have two options:
* change MathML Full to match MathML Core, and look for future patches to Firefox (& Safari)
* revert MathML Core to match MathML Full, and look for future patches to Chrome (& Safari)

Before pushing for either of those two options, it would be good to understand whether the  changed behavior specified by Core was intentional, and the reasoning behind it. It is hard for me to see the use-case; OTOH, in the current Full spec, percentage, dimensionless numbers and em units are somewhat redundant. (unitless was not carried over into Core)

**Comment by bkardell** [comment]
The WG discussed this yesterday, minutes follow

> 
> DG: a single toggle of how you intrepret a percentageof minsize and maxsize. 
> we want to align mathml 4 and the core. 
> 
> FW: I would need to check this one. I don't have a strong preference as a user. But as an implementer, just want to make sure we are not introducing something that violates browser's layout model or complicates the implementation needlessly (that might be the reason of the change, but not can't remember)? At glance, it seems quite easy to get calculate a percent of the target size while during layout for the other option we would have to calculate a percent of the unstretch size of a core operator (which can be deep in the tree due to embellished op and so on). Would be worth giving a try with Chromium...
> 
> NS: we need to be governed by what is implementable. It is possible that Fred made the same mistake that David and I made about thinking it was referencing the stretched size... or not.
> 
> BM: it could be CSS convention that causes the confusing.
> 
> DG: In CSS, max height refers to the parent.
> 
> NS: How much will break if we change the spec?
> 
> BM: In DLMF, specific sizes are not used much but growable sizes look better IMHO.
> 
> DG: ar5iv 2024 stats:
> mo: 4,704,693,527
> mo@maxsize: 69,820,857
> 
> 
> From Brian Kardell to Everyone: https://github.com/HTTPArchive/almanac.httparchive.org/discussions/3619 ... actually, maybe disregard, that wouldn't be good as a data source because it is biased away from stuff with math to begin with even.
> 
> bk: if you have further thoughts share them on the  issue and let fred how important his is.

**Comment by fred-wang** [comment]
Hi, so I get a chance to look this into more details.

For reference, current implementations are here:

Firefox: https://searchfox.org/mozilla-central/rev/b4860b54810539f0e4ab1fc46a3246daf2428439/layout/mathml/nsMathMLmoFrame.cpp#619

WebKit: https://searchfox.org/wubkat/rev/3180ee0bf2d7f95d1d938b55546a3ef50eb2b9b6/Source/WebCore/rendering/mathml/RenderMathMLOperator.cpp#124

Chromium: https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/layout/mathml/math_operator_layout_algorithm.cc;l=71;drc=18e5fcceb38d3cb42ba12379c08b698df9592876

First, I stand corrected about the complexity. All the implementations handle the calculation in the layout classes for the mo element. In the past, the minsize/maxsize resolution was handled in https://w3c.github.io/mathml-core/#algorithm-for-stretching-operators-along-the-block-axis but later I moved this to https://w3c.github.io/mathml-core/#layout-of-operators to simplify things.

So for this issue we can just replace "Percentage values are interpreted relative to T = Tascent + Tdescent" with "Percentage values are interpreted **relative to the unstretched size**" (maybe definining it more precisely as the height of the bounding box of glyph corresponding to c).

Bruce mentioned some links to MathML full and even if "normal" can be confusing, I believe the "unstreched size" is clearer. I can't remember why this was done that way in MathML Core, maybe because of this unclear "normal" wording or maybe because in the past this was done in another place where the unstretched size couldn't be easily retrieve.

From a user point of view, taking a percent of the target size also means targetting a different size which is probably not what we want? I can imagine the use of 105% to say make things slightly taller but 200% would mean introducing big gaps. I see that font units (e.g. 3em) allows to approximate multiple of the unstretched size (e.g. 300%) but probably one wants more accuracy.

Regarding the "symmetric" property, all browsers and MathML Core do that before applying the minsize/maxsize constraints. Applying "symmetric" can increase the size (multiplying it by a factor between one and two), so 1) if minsize would be applied before it could make the character unnecessarily big and 2) we can't apply the maxsize constraint before, without garanteeing it won't be violated later by symmetrization.

It also seems all browsers and current MathML Core spec are wrong, since they somehow trying to apply scaling using the baseline as a reference rather the math axis. Step 5 can just be corrected by a coordinate change :

- If T ≤ 0 then set Tascent to **minsize / 2 + AxisHeight** and then set Tdescent to minsize − Tascent.
- Otherwise, if 0 < T < minsize then **Tascent to (Tascent - AxisHeight) * minsize / T + AxisHeight** and Tdescent to minsize - Tascent. 
- Otherwise, if maxsize < T then **Tascent to (Tascent - AxisHeight) * maxsize / T + AxisHeight** and Tdescent to minsize - Tascent.

If symmetrization happened in step 4, then at step 5 we have:

Tascent − AxisHeight = Tdescent + AxisHeight = T/2

and after applied the minsize/maxsize constraints this equality remains true (T is respectively minsize, minsize and maxsize for the three bullets) so the operator remains symmetric with respect to the math axis.

Note that the math axis is generally small for normal math fonts so this difference is not really observable in practice. Still, we can write tests with math fonts that have large AxisHeight in order to test that more precisely.

**Comment by fred-wang** [comment]
# Chromium

![chromium-release](https://github.com/w3c/mathml-core/assets/567455/e6ff1844-4870-419a-b295-c25e19d7f312)

# Firefox
 ![firefox-release](https://github.com/w3c/mathml-core/assets/567455/0059dafd-40ac-4141-a205-103a1641e03e)

# WebKit (dev)

![webkit-dev](https://github.com/w3c/mathml-core/assets/567455/a3eb20c3-56dc-4a06-a2f1-277f67d37825)

# Chromium (dev + patch)

See https://chromium-review.googlesource.com/c/chromium/src/+/5402478/1/third_party/blink/renderer/core/layout/mathml/math_operator_layout_algorithm.cc 

![chromium-dev-patched](https://github.com/w3c/mathml-core/assets/567455/16e0ae42-fd57-4877-8902-e02cf392fbb9)

**Comment by brucemiller** [comment]
This looks fantastic! Thanks for digging into this @fred-wang !!

Personally, I can live with the subtle differences in the unsymmetric cases, since I'm never completely sure what I expect there anyway :>

**Comment by fred-wang** [comment]
Regarding the tests, the only one affected by the change is https://chromium-review.googlesource.com/c/chromium/src/+/5402478/2/third_party/blink/web_tests/external/wpt/mathml/presentation-markup/operators/mo-minsize-maxsize-001.html#118

I'm not sure we have a test to verify symmetric=true + minsize/maxsize in general, so will try to add one with the AxisHeight subtility.

**Comment by NSoiffer** [comment]
Sounds like we have a resolution. I've added the "needs specification update" label. From what @fred-wang  wrote, there are two changes:

1. So for this issue we can just replace "Percentage values are interpreted relative to T = Tascent + Tdescent" with "Percentage values are interpreted __relative to the unstretched size__" (maybe definining it more precisely as the height of the bounding box of glyph corresponding to c).

2. If T ≤ 0 then set Tascent to **minsize / 2 + AxisHeight** and then set Tdescent to minsize − Tascent.
   - Otherwise, if 0 < T < minsize then set **Tascent to (Tascent - AxisHeight) * minsize / T + AxisHeight** and Tdescent to minsize - Tascent. 
   - Otherwise, if maxsize < T then set **Tascent to (Tascent - AxisHeight) * maxsize / T + AxisHeight** and Tdescent to maxsize - Tascent.
\
  If symmetrization happened in step 4, then at step 5 we have:
\
  Tascent − AxisHeight = Tdescent + AxisHeight = T/2
\
  and after applied the minsize/maxsize constraints this equality remains true (T is respectively minsize, minsize and maxsize for the three bullets) so the operator remains symmetric with respect to the math axis.

**Comment by fred-wang** [comment]
Thanks. If there is no complaints, I will go ahead and work on spec/tests/impl as soon as I have time (hopefully later this week).

**Comment by fred-wang** [comment]
There is one note about the interpretation of maxsize ∞ that should be moved closer if possible, and I would also add a note about the fact that scaling for maxsize/minsize is done with respect to the math axis, hence preserving any previous symmetry from step 4.

**Comment by fred-wang** [comment]
I pushed a commit to apply the discussed changes above as well as more notes explaining these steps.

Note that MathML full was also scaling with respect to the baseline (rather than the math axis) so it was not respecting the symmetric rule either: "If the total size = height+depth is less than minsize or greater than maxsize, increase or decrease both height and depth proportionately so that the effective size meets the constraint. "

**Comment by fred-wang** [comment]
There seem to be another change that happened: **in MathML full minsize default to 100% while in MathML Core, it defaults to 1em**. But it's likely 1em is larger than 100% the unstretched size, so MathML Core would always do scaling adjustment to ensure target size is at least 1em. I guess this was related to this issue regarding interpretation of percentage, but now we fixed that we can **go back to a default minsize of 100%**. I haven't checked yet, but maybe this can explain things like https://github.com/w3c/mathml-core/issues/205#issuecomment-1828727460

**Another thing we might revise is the case `T ≤ 0`**. In practice this is really `T = 0` and happens when all non-stretchy siblings have zero height/depth or that we try to [determine unstretched sizes when the mrow only contains stretchy elements](https://w3c.github.io/mathml-core/#dfn-algorithm-for-stretching-operators-along-the-block-axis) as shown in the testcase below. The scaling adjustment cannot be performed (division by zero) and it was probably just an edge case (does not happen in practice) so the default choice to set symmetric minsize/2 above and below the baseline (math axis with the latest fix) is fine. However, Firefox/WebKit don't seem to do that when the operator does not have the symmetric property.

I'm thinking we could instead to do something like this:

1. If the operator has the symmetric property then continue to set the Tascent
    to minsize / 2 + AxisHeight and Tdescent to minsize − Tascent.
2. **Otherwise (if the operator has the symmetric property), set Tascent / Tdescent to the ascent/descent of the unstretched glyph, scaled proportionately to meet the minsize i.e. Tascent = glyphascent * minsize / glyphheight and Tdescent = minsize - Tascent.**

In particular, when minsize is the default 100% and the operator does not have the symmetric property, then Tascent/Tdescent would just be the ascent/descent of the unstretched glyphs, so that unstretched glyph can just be used to render the stretched operator. Also Δ = 0 in step 6 and the unstretched glyph preserves its vertical alignment with respect to the baseline.

That said, I'm not even sure about the desired vertical alignment for operators and stretched glyph selection, even when there is only one non-stretchy element. See https://github.com/w3c/mathml-core/issues/205#issuecomment-2034350739 for a testcase.

```
<!DOCTYPE html>
<link
  rel="stylesheet"
  href="https://fred-wang.github.io/MathFonts/LatinModern/mathfonts.css" />
<meta charset="UTF-8">
<style>
  math { font-size: 32pt; }
</style>
<!DOCTYPE html>
<link
  rel="stylesheet"
  href="https://fred-wang.github.io/MathFonts/LatinModern/mathfonts.css" />
<meta charset="UTF-8">
<style>
  math { font-size: 32pt; }
</style>
<math display="block">
  <mrow>
    <mo symmetric="false">|</mo>
    <mpadded height="0" depth="0"><mn></mn></mpadded>
  </mrow>
  <mn>__−</mn>
  <mrow>
    <mo symmetric="false">(</mo>
    <mo symmetric="false">)</mo>
  </mrow>
</math>
<math display="block">
  <mrow>
    <mo minsize="5em" symmetric="false">|</mo>
    <mpadded height="0" depth="0"><mn></mn></mpadded>
  </mrow>
  <mn>__−</mn>
  <mrow>
    <mo minsize="5em" symmetric="false">(</mo>
    <mo minsize="5em" symmetric="false">)</mo>
  </mrow>
</math>
```

**Comment by fred-wang** [comment]
https://chromium-review.googlesource.com/c/chromium/src/+/5402478 has been updated to match the current spec and increase test coverage.

**Comment by fred-wang** [comment]
Tests updated/added in https://github.com/web-platform-tests/wpt/pull/45572

and the spec is updated too, so closing this.
